lang/ruby/lib/avro/schema.rb

```diff
@@ -485,6 +490,19 @@ def to_avro(names=nil)
         avro['scale'] = scale if scale
         avro
       end
+
+      def match_schema?(schema)
+        if type_sym == schema.type_sym
+          return false if logical_type != schema.logical_type
+
+          if logical_type == 'decimal'.freeze
```

Comment:
You should define this as a class-level (frozen) constant and then use `DECIMAL` here

Reply:
Sure, I can make a change like that, but do you know if actually makes a difference since the Ruby VM will treat this frozen string like a constant, and will not allocate a new one on each invocation:

```
irb(main):001:0> def frozen_string_alloc
irb(main):002:1>   "foo".freeze.object_id
irb(main):003:1> end
=> :frozen_string_alloc
irb(main):004:0> frozen_string_alloc
=> 70328047113620
irb(main):005:0> frozen_string_alloc
=> 70328047113620 # <---------------------- same object
```

Reply:
The above behavior depends on a few things:

- Strings shorter than a certain length are allocated differently than strings over that length. I think it's 19 bytes, but don't quote me.
- Newer Ruby versions (and irb) may freeze strings by default. Depends on your distro's config and whatnot. The `# frozen_string_literal: true` pragma in a .rb file will force this.
- `"foo"` might allocate a string, but `"foo".freeze` looks it up in the string table, resulting in the same object ID as you would expect. But you still allocated `"foo"`. You'll have to use something like `memory_profiler` to know for sure (and this will work differently in say Ruby 2.4 than it does in Ruby 2.7).
- Ruby applications which are compiled into `iseq` files before executing (e.g., with Bootsnap) might result in different behavior in terms of whether the original non-frozen string is allocated each time the method is run or not.

One thing to note as well is that Ruby constants don't necessarily work like you would expect. `FOO = String("bar")` is legal, and then you can `FOO.concat("baz")`. In Ruby versions prior to them implementing immutable strings (so, IIRC, 2.5 and earlier), `"bar"` and `String("bar")` are the same. You can freeze the string, which makes `.concat` illegal, but constant values aren't frozen by default.

Basically what I want to point out is that you might be doing something that is interpreted like this:

```
every call:
  allocate string "foo"
  lookup this string in the string table
  return frozen reference
  deallocate string "foo"
```

but what you actually want is:

```
on load:
  allocate string "foo"
  lookup and store string in the string table

every call:
  get frozen reference
```

The latter should be the case with `# frozen_string_literal: true`, but this library targets older Ruby versions than that and so you probably want to ensure that happens.

Reply:
Thanks @ziggythehamster, I had not considered that a string may be allocated before the frozen reference is looked up!